sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/DoFnReflector.java

```diff
@@ -95,1056 +39,288 @@
 import java.util.LinkedHashSet;
 import java.util.List;
 import java.util.Map;
-import javax.annotation.Nonnull;
-import javax.annotation.Nullable;
-
 
 /**
- * Utility implementing the necessary reflection for working with {@link DoFn}s.
+ * Parses a {@link DoFn} and computes its {@link DoFnSignature}. See {@link #getOrParseSignature}.
  */
 public abstract class DoFnReflector {
+  private DoFnReflector() {}
 
-  private static final String FN_DELEGATE_FIELD_NAME = "delegate";
-
-  private enum Availability {
-    /** Indicates parameters only available in {@code @ProcessElement} methods. */
-    PROCESS_ELEMENT_ONLY,
-    /** Indicates parameters available in all methods. */
-    EVERYWHERE
-  }
-
-  /**
-   * Enumeration of the parameters available from the {@link ExtraContextFactory} to use as
-   * additional parameters for {@link DoFn} methods.
-   * <p>
-   * We don't rely on looking for properly annotated methods within {@link ExtraContextFactory}
-   * because erasure would make it impossible to completely fill in the type token for context
-   * parameters that depend on the input/output type.
-   */
-  private enum AdditionalParameter {
-
-    /** Any {@link BoundedWindow} parameter is populated by the window of the current element. */
-    WINDOW_OF_ELEMENT(Availability.PROCESS_ELEMENT_ONLY, BoundedWindow.class, "window") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return TypeToken.of(BoundedWindow.class);
-      }
-    },
-
-    INPUT_PROVIDER(Availability.PROCESS_ELEMENT_ONLY, DoFn.InputProvider.class, "inputProvider") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return new TypeToken<DoFn.InputProvider<InputT>>() {}.where(
-            new TypeParameter<InputT>() {}, in);
-      }
-
-      @Override
-      public boolean isHidden() {
-        return true;
-      }
-    },
-
-    OUTPUT_RECEIVER(
-        Availability.PROCESS_ELEMENT_ONLY, DoFn.OutputReceiver.class, "outputReceiver") {
-      @Override
-      public <InputT, OutputT> TypeToken<?> tokenFor(TypeToken<InputT> in, TypeToken<OutputT> out) {
-        return new TypeToken<DoFn.OutputReceiver<OutputT>>() {}.where(
-            new TypeParameter<OutputT>() {}, out);
-      }
+  private static final Map<Class<?>, DoFnSignature> SIGNATURE_CACHE = new LinkedHashMap<>();
```

Comment:
(That would also make it feasible to mock this, which may be useful in future tests, etc.)